### PR TITLE
bpo-45081: Fix __init__ method generation when inheriting from Protocol

### PR DESCRIPTION
**Pull request by uriyyo**
https://bugs.python.org/issue45081

### CODE REVIEW

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please verify that we need backport those changes to 3.9 and 3.10 versions?

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -1015,6 +1015,11 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
         # Does this class have a post-init function?
         has_post_init = hasattr(cls, _POST_INIT_NAME)
 
+        # typing.Protocol can override __init__ method to object.__init__
+        inherits_from_protocol = any(getattr(c, '_is_protocol', False) for c in cls.__bases__)
+        if inherits_from_protocol and cls.__dict__.get('__init__') is object.__init__:
+            del cls.__init__
```
Comment (Fidget-Spinner) [inline Lib/dataclasses.py]: A question: will this always succeed? What if `cls` is an immutable type? Or is that not possible?

File: Lib/dataclasses.py [hunk 2]
```diff
@@ -1015,6 +1015,11 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
...
+
```
Comment (ericvsmith) [inline Lib/dataclasses.py]: Could you improve this comment to more fully explain what's going on? I don't understand the issue by looking at this. What is it special about typing.Protocol that's different from any other class's implementation of __init__ that requires it to be special cased?

**Review by Fidget-Spinner — COMMENTED** [review]
@uriyyo if I understand the original issue correctly, you're right that we need to backport this to fix that issue's backports.

I'm deferring this to @serhiy-storchaka or @ambv because I'm not too familiar with the previous patch that caused this regression. Sorry.

File: Lib/dataclasses.py (hunk 1 above)
Comment (uriyyo) [reply Lib/dataclasses.py]: I think it should always succeed as far as `typing.Protocol` doing similar things with classes. `if` body will be executed only for a `typing.Protocol` subclasses so there should be no errors.

File: Lib/dataclasses.py (hunk 2 above)
Comment (uriyyo) [reply Lib/dataclasses.py]: Looks like I found a way how to fix this issue without changing `dataclasses` module.

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please review this PR? Looks like I found way how to fix this issue without changing `dataclasses` module.

File: Lib/typing.py [hunk 3]
```diff
@@ -1401,7 +1401,25 @@ def _is_callable_members_only(cls):
 
 
 def _no_init(self, *args, **kwargs):
-    raise TypeError('Protocols cannot be instantiated')
+    cls = type(self)
+
+    if cls._is_protocol:
+        raise TypeError('Protocols cannot be instantiated')
```
Comment (Fidget-Spinner) [inline Lib/typing.py:1407]: Note to self: this is the previous behavior.

File: Lib/typing.py [hunk 4]
```diff
@@ -1401,7 +1401,25 @@ def _is_callable_members_only(cls):
...
+
+    # set correct __init__ method on a first initialization
```
Comment (Fidget-Spinner) [inline Lib/typing.py]: Wow the fix is really smart!
I was kind of confused, so could you please change the comments to this? (After renaming `_no_init` to something better, see the other comment).
```
    # Initially, `__init__` of a protocol subclass is set to `_no_init`.
    # The first instantiation of the subclass will call `_no_init` which
    # searches for a proper new `__init__` in the MRO. The new `__init__`
    # replaces the subclass' old `__init__` (ie `_no_init`). Subsequent
    # instantiation of the protocol subclass will thus use the new
    # `__init__` and no longer call `_no_init`.
```

File: Lib/typing.py [hunk 5]
```diff
@@ -1401,7 +1401,25 @@ def _is_callable_members_only(cls):
 
 
 def _no_init(self, *args, **kwargs):
```
Comment (Fidget-Spinner) [inline Lib/typing.py]: Please rename this to something like `_no_init_or_replace_init` (I admit my proposed name is terrible, but I couldn't think of a better one. If you think of something better, please use that instead.)

File: Lib/test/test_dataclasses.py [hunk 6]
```diff
@@ -2150,6 +2150,25 @@ def __init__(self, x):
                 self.x = 2 * x
         self.assertEqual(C(5).x, 10)
 
+    def test_inherit_from_protocol(self):
+        # See bpo-45081.
```
Comment (Fidget-Spinner) [inline Lib/test/test_dataclasses.py:2155]: To save future generations some time. Please change this to
```suggestion
        # Dataclasses inheriting from protocol should preserve their own `__init__`.
        # See bpo-45081.
```

**Review by Fidget-Spinner — COMMENTED** [review]
After researching this issue (and the previous patch), this LGTM in general, I only have suggestions to change wording in some parts.

I'm not very confident with backporting this change so late into 3.10's release cycle. I hope another core dev can review.

File: Lib/typing.py (hunk 5 above)
Comment (uriyyo) [reply Lib/typing.py]: I am thinking about something more general like `_protocol_init`

File: Lib/typing.py (hunk 4 above)
Comment (uriyyo) [reply Lib/typing.py]: Thanks for such a good comment, it will definitely make things more clean for other maintainers.

File: Lib/typing.py (hunk 5 above)
Comment (Fidget-Spinner) [reply Lib/typing.py]: That works too :). I'll leave it up to you.

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. I will wait a bit to see if Serhiy or Lukasz (or anyone else) says anything.

**Comment by ambv** [comment]
The previous behavior was introduced in GH-27545 as an attempt to make non-protocol subclasses of Protocol ignore the `__init__` inherited from protocol base classes.

This PR retains this ability and restores compatibility with `dataclasses`. LGTM.

**Pull request merged**